File: Alexander Chibirev/lab7/lab7/lab7/BaseSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab7
{
    abstract class BaseSort : ISortCalculator
    {

        protected BaseSort()
        {
            List<int> vec = new List<int>();
            m_array = vec;
        }

        public List<int> GetResult()
        {
            return m_array;
        }

        public void Reset(List<int> array)
        {
            m_array = array;
        }

        public void Sort(bool isParallel)
        {
            m_isParallelEnabled = isParallel;
            SortArray();
        }

        protected abstract void SortArray();

        protected List<int> m_array;
        protected bool m_isParallelEnabled = false;
    }
}

File: Chibirev Alexander/lab6/lab6/lab6/BaseSort.cpp
#include "stdafx.h"
#include "BaseSort.h"

CBaseSort::CBaseSort(const std::vector<int>& vec)
    : m_array(vec)
    , m_isParallelEnabled(false)
{
}

void CBaseSort::WaitForMultiplyObjects(std::vector<std::thread>& threads)
{
    std::for_each(threads.begin(), threads.end(), [](std::thread& thread) { thread.join(); });
}


void CBaseSort::Sort(bool isParallel)
{
    m_isParallelEnabled = isParallel;
    SortArray();
}

std::vector<int> CBaseSort::GetResult() const
{
    return m_array;
}

void CBaseSort::Reset(const std::vector<int>& array)
{
    m_array = array;
}